### ISSUE TITLE
Reactive programming with Spring 5 for JHipster

### ISSUE THREAD

**Issue by JulienMrgrd** [jhipster/generator-jhipster#5970]
Hi! 
We are 2 interns (Joffrey Derchu @JoffreyDcu and Julien Margarido @JulienMrgrd) working on JHipster with @Jdubois, at the Ippon office in Paris.
 
As you may know, Spring5 is about to be released. It will introduce many [new features ](https://github.com/spring-projects/spring-framework/wiki/What%27s-New-in-the-Spring-Framework) including support for reactive programming. 
We are looking to make efficiency and stress tests on applications using Spring5’s reactive programming interface, in order to get measurable results for performance, consistency and scalability.
The applications will be generated by JHipster with various configurations (relational or non-relational databases, cache, Kafka, ElasticSearch,…) and then migrated to Spring5 (with Webflux module) and Spring Boot 2.

   Our goal is to determine in which cases using Spring5’ WebFlux module is a good idea. If the results are interesting, we will consider adding support for reactive programming on the generator (maybe new questions?).
 
For your information, we already have an application running with Spring5/SB2. We are currently coding REST methods using WebFlux (Mono and Flux). 
Our repo : https://github.com/JulienMrgrd/webflux-jhipster.

If you have any questions, remarks or ideas, please add a comment here.

### <b>EDIT :</b> Here is our article : http://blog.ippon.tech/spring-5-webflux-performance-tests/

<br>

_A few links_ : 

- The Reactive Manifesto: https://github.com/reactivemanifesto/reactivemanifesto
- Spring docs : http://docs.spring.io/spring/docs/5.0.0.BUILD-SNAPSHOT/spring-framework-reference/htmlsingle/#web-reactive
http://docs.spring.io/spring/docs/5.0.x/spring-framework-reference/
- (video) Reactive Spring by Josh Long : https://www.youtube.com/watch?v=EDB24onZxsA

**Comment by vw98075** [comment]
That will be great! When is the targeted release date?

**Comment by nicoraynaud** [comment]
Thank you in advance for this work.

I look forward to see your conclusions on whether this adds value to the current architecture.

**Comment by marcelinobadin** [comment]
Awesome! I am starting a new project right now, once you release that I will immediately start using it and I can give you any feedback you need.

**Comment by vw98075** [comment]
I guess the starting point would be using reactive for all repository functions as well as all Rest end points. And the conversion would be expanded to other resource sections of JHipster components.

**Comment by gzsombor** [comment]
I'm curious how do you want to proceed, and what are you expecting, as Hibernate and the JDBC connections are still working in a blocking way, and based on the problems with  'LazyInitializationException' there are lot's of projects, where the transaction are not handled properly - or not understood et all.

**Comment by jdubois** [comment]
@gzsombor I shared my plan for "JHipster Webflux" on the core dev mailing list, and for the record I totally agree with you.

Anyway the goal is here is to test and measure everything. I don't believe the conference hype until I see numbers.

**Comment by Tcharl** [comment]
Hi,

Nice study: as I said, true hipsters want reactive programming ;-).
Is the purpose of this thread to grab so business case where reactive programming could help?
If so, I had in mind that it could replace the 'event/command' pattern described in this [post](https://www.infoq.com/articles/microservice-event-choreographies?utm_source=sumome&utm_medium=email&utm_campaign=sumome_share) by the use of long lived streams.
Also, the big data ecosystem (using kafka and so on) could inverse the paradigm from the 'push to lake' to 'continuous stream from source'. I may be wrong on my interpretation of reactive programing, and so the benefits could only consists in getting non-blocking end to end communication to avoid blocking ui and being lazy as we can, but getting rid of any event-driven paradigm could be a very nice way to go, as this paradigm is usually very hard to tackle.

**Comment by JulienMrgrd** [comment]
Here is our article : http://blog.ippon.tech/spring-5-webflux-performance-tests/

**Comment by atomfrede** [comment]
Nice article! We implemented a reactive gateway with vert.x recently and also did not find some increase in performance. But one node could handle a **lot** more requests compared to a classical architecture. Have you tried to find the point when the classic stack breaks under heavy load? If everything is reactive the reactive application should be able to handle more load.

Another interesting observation we did with vert.x was, that while the classical approach just stops working when load is to high the reactive counterpart only becomes veerryy slow but answers the request finally. Did you do some experiments on how spring web flux behaves here?

**Comment by gmarziou** [comment]
There's also the |Spring Cloud Gateway](https://github.com/spring-cloud/spring-cloud-gateway)

**Comment by raharrison** [comment]
Very interesting. It seems that the reactive approach with Spring definitely isn't ready for prime-time just yet.

**Comment by cbornet** [comment]
It would be interesting to test a streaming use case (that's where reactive shines IMO)

**Comment by nkolosnjaji** [comment]
@JulienMrgrd I think this applications are not fully reactive as DomainUserDetailsService is still blocking, right? Reactive application must not block, as it uses very little number of threads.

@cbornet maybe streaming memory from metrics endpoint would be nice, so user can observe memory in real-time during load-test or production in a plot.

**Comment by nkolosnjaji** [comment]
Mongo async driver is using unlimited number of threads so I guess it creates a lot of threads during load test.
See [more](https://jira.mongodb.org/browse/JAVA-2561)

**Comment by cbornet** [comment]
Just wanted to share that if you want reactive templates for JHipster's endpoints, you can check the [grpc module](https://github.com/cbornet/generator-jhipster-grpc/blob/master/generators/app/templates/_AccountService.java). It uses RxJava2 but it should be easy to convert to Reactor since the APIs are very similar.

**Comment by deepu105** [comment]
@jduboisI tried the module from @cbornet it's awesome. Maybe we should make it the official module for reactive programming as it provides reactive endpoint for all our REST endpoints including entities. 
@cbornet how difficult is it to make the module spring5/jhipster 5 compatible?

**Comment by deepu105** [comment]
ping @jdubois

**Comment by jdubois** [comment]
@deepu105 @cbornet I wanted to add Spring Webflux support next week, so it's the perfect time to discuss this.

- Does it make sense to have the whole application with reactive APIs? As we saw it wasn't doing much good when using a database (even MongoDB!), so my idea was to only add this to the Spring Controller sub-generator. It's not a very much used sub-generator unfortunately, but that's where it makes more sense to me if you want to have reactive APIs
- I'm of course in favor of doing the less work possible, so if @cbornet is OK to have his code in the main project, that's very good for me.

**Comment by deepu105** [comment]
@jdubois I think you misunderstood me. I'm not talking about merging the module to the main generator. We shouldn't do that. Instead, we can move the module into jhipster org (if @cbornet is ok with that) and advertise that as the official solution for reactive and will become an official module/blueprint from us but still in a separate repo

**Comment by jdubois** [comment]
@deepu105 yes but reactive is *supposed* to be so important, we should have an official option inside the generator, not just a blueprint.

**Comment by cbornet** [comment]
Of course we can put the code into JHipster org ! It’s currently done for gRPC so there will be some little adaptations to do for Webflux (eg. for pagination). And I haven’t done entities without service (that’s on my TODO since a too long time). I’ll be glad to help with a Webflux blueprint.

**Comment by jdubois** [comment]
OK so what do we do @cbornet ? If you are in La Défense we can meet up and discuss this

**Comment by cbornet** [comment]
I’m not based in La Défense anymore :smile: . I’m now at St Lazare. Not at Google, but just next to it.

**Comment by jdubois** [comment]
@cbornet well that's also close to Ippon - I'll send you an email!

**Comment by jdubois** [comment]
As JHipster 5 is coming very soon, we need to tackle this issue.

First of all, great work @JoffreyDcu @JulienMrgrd - I ran through all you code again, and for me this is a very good analysis. Besides, the "official word" from the Spring team has changed, have a look at the [current documentation](https://docs.spring.io/spring/docs/current/spring-framework-reference/web-reactive.html) where it says (as of 02/24/2018):

`If you have blocking persistence APIs (JPA, JDBC), or networking APIs to use, then Spring MVC is the best choice for common architectures at least. It is technically feasible with both Reactor and RxJava to perform blocking calls on a separate thread but you wouldn’t be making the most of a non-blocking web stack.` - which is exactly what you did and what you found out.

Of course you can still find people on the Internet claiming you can have full reactive application with Spring Data JPA, like [in this article](https://thepracticaldeveloper.com/2017/11/04/full-reactive-stack-with-spring-webflux-and-angularjs/#Reactive_Repository_with_Spring_5_and_Spring_Data_JPA_20) - but if read it, you'll notice the author talks about Spring Data JPA but uses MongoDB, so it just looks like he has no idea what he's talking about...

There is of course the case of "reactive" databases like MongoDB, but unless you have a sharded cluster, you still have just one master node, and for me that's quite logical that the analysis didn't find a great performance gain here: we are locking writes on that node, and as writes are the biggest issue (at least with JHipster, as we have a strong emphasis on caching), this is normal that it doesn't work better than with Spring MVC.

Which leaves us with 4 use cases for reactive programming:

- MongoDB with a sharded cluster. This would need to be proven, but I agree this looks promising. Then, nobody does sharded clusters with MongoDB. I talked a few months ago to https://mlab.com which is probably the biggest MongoDB hosting services (and I also talked to several competitors), and all they offered at the time was vertical scaling. As of today you can see at https://mlab.com/plans/pricing/#plan-type=dedicated that there is sharded cluster at the very bottom (beware, there is also a "shared cluster", which is totally different!), but it's clear that it's not their most common option
- Cassandra works probably very well. But we have less than 1.5% of users with Cassandra. And also, I've already done some very efficient applications using their async driver + Spring MVC, I'm not sure anybody would need more (and given the size and price of our Cassandra ring, scaling horizontally the app servers wouldn't have been an issue anyway).
- Couchbase also probably works well. But here we're talking 0,1% of our users.
- People using "something else", like doing REST requests to a third-party system, like doing IoT. For me, this is where there is the biggest potential gain.

So I talked with @cbornet on Friday evening and he told me he could port his gRPC code to JHipster -> of course that's the approach I would prefer to have, if @cbornet you have the time.

Otherwise, I think that we could go with a very simple solution:

- Add a "Spring Webflux" option to the main generator
- Add the option to have Spring Webflux only to the Spring Controller sub-generator (which would handle my 4th case above)

-> that would be very easy to add, and be good enough for most people as I think only my 4th case would really be useful

**Comment by raharrison** [comment]
Maybe it's better to wait off until async jdbc finally gets here, at which point there would be a more clear use case for the majority of people? At the moment is it worth adding in and then supporting/updating it when I guess it's unlikely it would be used that much? (although true your 4th point is interesting use case).

That being said I feel like I've been waiting on async jdbc for years now, but it does look more promising this time around. I think it was initially targeted at jdk10, but I guess jdk11 is more likely now. We would also have to wait for driver support though before using.

http://hg.openjdk.java.net/jdk10/sandbox/jdk/file/a31057bda7c5/src/java.sql/share/classes/java/sql2

**Comment by cbornet** [comment]
Ok. What should we start with ? Entities or other endpoints ?

Also another great use case for Webflux should be the streaming of large collections since with reactive streams, the collection is not fetched fully into memory but progressively at the speed of the consumer.

**Comment by jdubois** [comment]
Spring Controller sub generator then entities then other endpoints.
Let's start by the small and obvious. The admin endpoints don't have much
load anyway.

Le 24 févr. 2018 19:09, "Christophe Bornet" <notifications@github.com> a
écrit :

**Comment by jdubois** [comment]
Oh and for entities, the MongoDB entities are probably the only important ones, so they also have the highest priority.
So I would go:

- Spring Controller sub-generator
- Entities with MongoDB
- Entities with JPA using something like the AsyncUtil class from @JulienMrgrd @JoffreyDcu -> https://github.com/jhipster/webflux-jhipster/blob/mongo-reactive/onlyoperationReactive/src/main/java/jhipster/reactive/web/rest/util/AsyncUtil.java (which should be in the JHipster server lib)
- And then if you have time some admin controllers, like login/logout, which are most used

-> no need to do Cassandra/Couchbase or less-used admin controllers, as we have too few users for those

**Comment by jdubois** [comment]
I just committed our first support, heavily inspired by your work @JulienMrgrd @JoffreyDcu !!

https://github.com/jhipster/generator-jhipster/commit/45ea0d13f527a310d59994b137476760f05d5c8d

After closer look, my next steps would be:

- Support for reactive with MongoDB entities (maybe even only support reactive for MongoDB??)
- Support for reactive in Spring Security (probably not such a big issue)

Not sure we can have this for JHipster 5.

**Comment by jdubois** [comment]
Closing this as I don't think we will have the time to do more for JHipster 5.

I also don't want to rush here:

- our tests show that "reactive" doesn't help that much with MongoDB. As always, MongoDB promises a lot but doesn't deliver as much as expected, and it gets flooded on its master node when you do lots of inserts. Like an SQL database, in fact. The only solution would be to use a sharded database, but very few people use this. So in the end reactive+mongodb is not such a great combination.
- for security: as most people now use a token-based approach, doing it in a separate thread doesn't really makes any sense, and probably costs more than you gain.

**Comment by mechero** [comment]
Hi @jdubois, I'm the author of [the article](https://thepracticaldeveloper.com/2017/11/04/full-reactive-stack-with-spring-webflux-and-angularjs) you referenced [above](https://github.com/jhipster/generator-jhipster/issues/5970#issuecomment-368241635). Thanks for the constructive comment... 😒maybe next time you can also give me feedback on the blog, Twitter or GitHub. 

It's indeed an errata in that paragraph since I'm using Spring Data MongoDB, not Spring Data JPA. It's fixed now.

**Comment by jdubois** [comment]
Thanks for correcting the article title and text @mechero - this article was misleading, I'm sorry to have hurt you about this, but people were sending this to me and it just didn't make any sense.

**Comment by Mahdihp** [comment]
Reactive programming with Spring 5 for JHipster
